src/Blocks/BlockCactus.h

```diff
@@ -77,31 +66,101 @@ class cBlockCactusHandler :
 		return true;
 	}
 
-	virtual void OnUpdate(cChunkInterface & cChunkInterface, cWorldInterface & a_WorldInterface, cBlockPluginInterface & a_PluginInterface, cChunk & a_Chunk, int a_RelX, int a_RelY, int a_RelZ) override
-	{
-		if (CanGrow(a_Chunk, a_RelX, a_RelY, a_RelZ) == paGrowth)
-		{
-			a_Chunk.GetWorld()->GrowCactus(a_RelX + a_Chunk.GetPosX() * cChunkDef::Width, a_RelY, a_RelZ + a_Chunk.GetPosZ() * cChunkDef::Width, 1);
-		}
-	}
+
+
+
 
 	virtual ColourID GetMapBaseColourID(NIBBLETYPE a_Meta) override
 	{
 		UNUSED(a_Meta);
 		return 7;
 	}
 
-protected:
 
-	virtual PlantAction CanGrow(cChunk & a_Chunk, int a_RelX, int a_RelY, int a_RelZ) override
+
+
+	virtual int Grow(cChunk & a_Chunk, Vector3i a_RelPos, int a_NumStages = 1) override
 	{
-		auto Action = paStay;
-		if (((a_RelY + 1) < cChunkDef::Height) && (a_Chunk.GetBlock(a_RelX, a_RelY + 1, a_RelZ) == E_BLOCK_AIR))
+		// Check the total height of the cacti blocks here:
+		int top = a_RelPos.y + 1;
+		while (
+			(top < cChunkDef::Height) &&
+			(a_Chunk.GetBlock({a_RelPos.x, top, a_RelPos.z}) == E_BLOCK_CACTUS)
+		)
+		{
+			++top;
+		}
+		int bottom = a_RelPos.y - 1;
+		while (
+			(bottom > 0) &&
+			(a_Chunk.GetBlock({a_RelPos.x, bottom, a_RelPos.z}) == E_BLOCK_CACTUS)
+		)
+		{
+			--bottom;
+		}
+
+		// Refuse if already too high:
+		auto numToGrow = std::min(a_NumStages, a_Chunk.GetWorld()->GetMaxCactusHeight() + 1 - (top - bottom));
+		if (numToGrow <= 0)
 		{
-			Action = super::CanGrow(a_Chunk, a_RelX, a_RelY, a_RelZ);
+			return 0;
 		}
 
-		return Action;
+		BLOCKTYPE blockType;
+		for (int i = 0; i < numToGrow; ++i)
+		{
+			Vector3i pos(a_RelPos.x, top + i, a_RelPos.z);
+			if (!a_Chunk.UnboundedRelGetBlockType(pos, blockType) || (blockType != E_BLOCK_AIR))
+			{
+				// Cannot grow there
+				return i;
+			}
+
+			a_Chunk.UnboundedRelFastSetBlock(pos, E_BLOCK_CACTUS, 0);
+
+			// Check surroundings. Cacti may ONLY be surrounded by non-solid blocks; if they aren't, drop as pickup and bail out the growing
+			static const Vector3i neighborOffsets[] =
```

Comment:
The cactus popping off seems perfect for a `OnUpdate`. Does the current code still work when I place a block next to the cactus?

Reply:
Should still work, thanks to `CanBeAt` (can't try right now). But it's true that the checking code should be reused, rather than having a copy here. Will fix.

Reply:
Can't fix easily, `CanBeAt` takes too different a set of parameters; changing `CanBeAt` to accept a single `cChunk` and `Vector3i` makes too many unrelated changes in other handlers, I'd prefer not to go that deep within this PR - especially when the block handlers will most likely be dropped entirely for the 1.13 change and then re-written from scratch.